arbornml/xmlwrap.hpp

```diff
@@ -0,0 +1,310 @@
+#pragma once
+
+// RAII and iterator wrappers for some libxml2 objects.
+
+#include <any>
+#include <cstddef>
+#include <cstdlib>
+#include <memory>
+#include <optional>
+#include <string>
+#include <utility>
+
+#include <libxml/parser.h>
+#include <libxml/xpath.h>
+#include <libxml/xpathInternals.h>
+
+#include <arbornml/nmlexcept.hpp>
+
+namespace arbnml {
+
+using non_negative = unsigned long long;
```

Comment:
It's a concept (more or less) defined by the schema, and it literally just means an arbitrary large non-negative integer.

It could be faithfully represented by a string, or by an arbitrary size bigint; `unsigned long long` is simply used to ease the implementation (and it seems to be a fairly common sort of short cut taken in XML-reading libs and applications).

Reply:
Using e.g. `std::uint64_t` means we still have a parsing problem to solve in the general case; indeed, it's not guaranteed to even be supported, though I'd be surprised if we have much trouble on modern machines.

Reply:
In short, I'd stick with `unsigned long long` or make it actually faithful and use a string representation.

Reply:
Great, then a comment to that effect will suffice ;)